### ISSUE TITLE
Overriding maps

### ISSUE THREAD

**Issue by fdncred** [kaby76/Piggy#10]
Now that things are compiling for Leptonica in Piggy 1.0.8 I have another question, of course. (Great work, BTW!!!)

How do I override the parameter and return maps that you moved to ClangSupport.pig? i.e.
```cs
public static Dictionary<string, string> _parm_type_map =
    new Dictionary<string, string>() {
    { "const char **", "out IntPtr" },
    { "char *", "[Out] byte[]"},
    { "unsigned int *", "out uint" },
    { "void **", "out IntPtr" },
    { "void *", "IntPtr" },
    { "const char *", "string" },
    { "const void *", "IntPtr" },
    { "const <type> *", "in <type>"},
};
```
Do I need to copy these files from the nuget Templates folder to my own and modify or what?

The reason I ask is because I don't like some of the default maps. As we've discussed, you'll eventually change "const char *" to a "string" but beyond that, leptonica has it's own defined types. i.e.
```c
/*--------------------------------------------------------------------*
 *                          Built-in types                            *
 *--------------------------------------------------------------------*/
typedef int                     l_ok;    /*!< return type 0 if OK, 1 on error */
typedef signed char             l_int8;     /*!< signed 8-bit value */
typedef unsigned char           l_uint8;    /*!< unsigned 8-bit value */
typedef short                   l_int16;    /*!< signed 16-bit value */
typedef unsigned short          l_uint16;   /*!< unsigned 16-bit value */
typedef int                     l_int32;    /*!< signed 32-bit value */
typedef unsigned int            l_uint32;   /*!< unsigned 32-bit value */
typedef float                   l_float32;  /*!< 32-bit floating point value */
typedef double                  l_float64;  /*!< 64-bit floating point value */
#ifdef COMPILER_MSVC
typedef __int64                 l_int64;    /*!< signed 64-bit value */
typedef unsigned __int64        l_uint64;   /*!< unsigned 64-bit value */
#else
typedef long long               l_int64;    /*!< signed 64-bit value */
typedef unsigned long long      l_uint64;   /*!< unsigned 64-bit value */
#endif  /* COMPILER_MSVC */
```
Since they're typedefs, it takes these and makes a struct out of them.

It seems to me that these could be defined in the map and be converted.  Something like this.
```cs
{ "l_ok", "int" },
{ "signed char", "byte" },
{ "unsigned char", "byte" },
// and so on
```
And then perhaps it wouldn't create a struct for l_ok, and the rest.

**Comment by kaby76** [comment]
We'll have to figure out something better. But for now, there's the 
ill-named method ModParamUsageType(Dictionary<string, string> 
additions). Declare in something like this:

template Project1Namespace : Namespace
{
     init {{
         namespace_name = "Csharp";

ClangSupport.ModParaUsageType(new Dictionary<string, string>() {

             { "size_t", "foofoo" },
             { "int", "barbar"}, ...
});

     }}
}

It'll override everything you give in that new dictionary. -K

**Comment by fdncred** [comment]
For documentation purposes, I made these changes and it seems to have worked. However it still creates the structs. I can manually delete the structs for items like `l_int32`. I just figured it would know, that since there is a map, not to create a struct for it.

```cs
template Project1Namespace : Namespace
{
	init {{
		namespace_name = "Native";

        ClangSupport.ModParamUsageType(new Dictionary<string, string>() {
            { "const char **", "out IntPtr" },
            { "char *", "[Out] byte[]"},
            { "unsigned int *", "out uint" },
            { "void **", "out IntPtr" },
            { "void *", "IntPtr" },
            { "const char *", "string" },
            { "const void *", "IntPtr" },
            { "const <type> *", "in <type>"},
            { "l_ok", "int" },
            { "l_int8", "byte" },
            { "l_uint8", "byte" },
            { "l_int16", "short" },
            { "l_uint16", "ushort" },
            { "l_int32", "int" },
            { "l_uint32", "uint" },
            { "l_float32", "float" },
            { "l_float64", "double" },
            { "l_int64", "long" },
            { "l_uint64", "ulong" },
            });

        ClangSupport.ModNonParamUsageType(new Dictionary<string, string>() {
            { "size_t", "SizeT" },
            { "int", "int"},
            { "uint", "uint"},
            { "short", "short"},
            { "ushort", "ushort"},
            { "long", "long"},
            { "unsigned char", "byte" },
            { "unsigned short", "UInt16"},
            { "unsigned int", "uint"},
            { "unsigned long", "ulong"},
            { "unsigned long long", "ulong"},
            { "long long", "long"},
            { "float", "float"},
            { "double", "double"},
            { "bool", "bool"},
            { "char", "byte"},
            { "const char *", "IntPtr" }, // For now, don't do [return: MarshalAs(UnmanagedType.LPStr)] set up of function.
            { "char *", "IntPtr" },
            { "signed char", "sbyte" },
            { "l_ok", "int" },
            { "l_int8", "byte" },
            { "l_uint8", "byte" },
            { "l_int16", "short" },
            { "l_uint16", "ushort" },
            { "l_int32", "int" },
            { "l_uint32", "uint" },
            { "l_float32", "float" },
            { "l_float64", "double" },
            { "l_int64", "long" },
            { "l_uint64", "ulong" },
            });
	}}
}
```

**Comment by kaby76** [comment]
The type mapping in functions isn't connected in any way to the struct 
pattern matching, but should be fixed. I have to think about this one 
because the struct pattern shouldn't even match. Hmm. "To do" today. -K

**Comment by kaby76** [comment]
A change to fix this is checked in for v1.0.9.
